### PR TITLE
Very large refactor of localkube

### PR DESCRIPTION
**Pull request by luxas**
Fixed lots of bugs, added flags and commands via cobra, extracted a simple interface for SimpleServer for all Kubernetes components, made DNS working again, created the type LocalkubeServer which holds all necessary data etc.

The job isn't done, but it's a step in the right direction.

Localkube requires the certs to be present to work right now, but that's the next step for us to do: get certs generated by localkube.

PTAL @dlorenc @ethernetdan @mfburnett

### CODE REVIEW

**Comment by luxas** [comment]
Note that the CI isn't affected as it's using a "stable" version of localkube.
We have to fix that too.

**Comment by dlorenc** [comment]
Nice! I'll take a look this evening. I'm working on a PR to start using the head version of localkube in minikube right now. Once these both land we'll be in really good shape.

**Comment by luxas** [comment]
New commits today:
 - IP flags are now "natively" `net.IP` and `net.IPNet`, which adds validation (no more `--service-cluster-ip-range=absbs`)
 - Certs are stored under `LocalkubeDirectory`
 - We're getting the public IP of the host (required for the DNS endpoint) in the same manner as apiserver creates the `kubernetes.default` endpoint => much more stable method than the previous
 - Remove the dep on `easyrsa`, localkube now signs its own certs
 - Add `ca.crt` to the ServiceAccounts. Absolutely required for InClusterConfig to function
 - Remove easyrsa code from `commands.go`

Now a really nice, standalone version of localkube with no deps starts to show up! 
This is great!

File: pkg/localkube/localkube.go [hunk 1]
```diff
@@ -71,4 +80,23 @@ func (lk LocalkubeServer) GetAPIServerInsecureURL() string {
...
+
+	alternateIPs := []net.IP{lk.ServiceClusterIPRange.IP}
+	alternateDNS := []string{fmt.Sprintf("%s.%s", "kubernetes.default.svc", lk.DNSDomain), "kubernetes.default.svc", "kubernetes.default", "kubernetes"}
+	hostIP, err := lk.GetHostIP()
+	if err != nil {
+		fmt.Println("Failed to get host IP: ", err)
+	}
+
+	if err := utilcrypto.GenerateSelfSignedCert(hostIP.String(), lk.GetPublicKeyCertPath(), lk.GetPrivateKeyCertPath(), alternateIPs, alternateDNS); err != nil {
+		fmt.Println("Failed to create certs: ", err)
```
Comment (dlorenc) [inline pkg/localkube/localkube.go]: Should this return the error so it can be handled above?

File: pkg/localkube/localkube.go [hunk 2]
```diff
@@ -17,61 +17,86 @@ limitations under the License.
...
-		fmt.Fprintln(out, "LocalKube Status")
-		fmt.Fprintln(out, "################\n")
-
-		fmt.Fprintln(out, "Order\tStatus\tName")
-		for num, server := range lk.Servers {
-			fmt.Fprintf(out, "%d\t%s\t%s\n", num, server.Status(), server.Name())
-		}
-
-		fmt.Fprintln(out)
+	if err := utilcrypto.GenerateSelfSignedCert(hostIP.String(), lk.GetPublicKeyCertPath(), lk.GetPrivateKeyCertPath(), alternateIPs, alternateDNS); err != nil {
```
Comment (dlorenc) [inline pkg/localkube/localkube.go]: Awesome! I didn't realize this existed.

Comment (luxas) [inline pkg/localkube/localkube.go]: I searched a while... and there it was! I knew apiserver did this automatically, but if we relied on that, there would be a race condition between apiserver and controller-manager, so this approach is better

File: pkg/localkube/localkube.go (hunk 1 above)
Comment (luxas) [inline pkg/localkube/localkube.go]: The whole error handling have to be refactored and taken care of as a whole. 
The same for logging.
But updating this to return the error :)

**Comment by luxas** [comment]
Updated

File: pkg/localkube/localkube.go [hunk 3]
```diff
@@ -17,61 +17,90 @@ limitations under the License.
...
+	alternateIPs := []net.IP{lk.ServiceClusterIPRange.IP}
+	alternateDNS := []string{fmt.Sprintf("%s.%s", "kubernetes.default.svc", lk.DNSDomain), "kubernetes.default.svc", "kubernetes.default", "kubernetes"}
+	hostIP, err := lk.GetHostIP()
+	if err != nil {
+		fmt.Println("Failed to get host IP: ", err)
+		return err
 	}
-	return errors.New("you must choose start <name>, stop <name>, or status")
+
+	if err := utilcrypto.GenerateSelfSignedCert(hostIP.String(), lk.GetPublicKeyCertPath(), lk.GetPrivateKeyCertPath(), alternateIPs, alternateDNS); err != nil {
```
Comment (ethernetdan) [inline pkg/localkube/localkube.go]: Will this be an issue if the VM gets assigned a different IP in the future?

Comment (dlorenc) [inline pkg/localkube/localkube.go]: +1, we should probably regenerate if the IP changes or if the file doesn't exist yet. After that this should be good to merge.

**Comment by dlorenc** [comment]
I'll just merge now. I got a branch working with the IP checking here:
https://github.com/dlorenc/minikube/tree/remove_weave

I'll merge that after.

File: pkg/localkube/localkube.go [hunk 4]
```diff
@@ -17,61 +17,90 @@ limitations under the License.
...
+	return fmt.Sprintf("https://%s:%d", lk.APIServerAddress.String(), lk.APIServerPort)
+}
+
+func (lk LocalkubeServer) GetAPIServerInsecureURL() string {
+	return fmt.Sprintf("http://%s:%d", lk.APIServerInsecureAddress.String(), lk.APIServerInsecurePort)
+}
+
+// Get the host's public IP address
+func (lk LocalkubeServer) GetHostIP() (net.IP, error) {
+	return utilnet.ChooseBindAddress(net.ParseIP("0.0.0.0"))
```
Comment (dlorenc) [inline pkg/localkube/localkube.go]: Whoops, this is choosing the wrong address for me. It looks like the code chooses the IP that is attached to a gateway, which is wrong on my VM:

$ cat /proc/net.route
Iface	Destination	Gateway 	Flags	RefCnt	Use	Metric	Mask		MTU	Window	IRTT
eth0	00000000	0202000A	0003	0	0	1	00000000	0	0	0
eth0	0002000A	00000000	0001	0	0	0	00FFFFFF	0	0	0
lo	0100007F	00000000	0005	0	0	0	FFFFFFFF	0	0	0
docker0	000011AC	00000000	0001	0	0	0	0000FFFF	0	0	0
eth1	0063A8C0	00000000	0001	0	0	0	00FFFFFF	0	0	0

This code chooses eth0, which has IP 10.0.2.15. etho1 has 192.168.99.100.

We might have to change this back to your original method, of looking for 192.168.

Comment (luxas) [inline pkg/localkube/localkube.go]: Strange, it's using the same method as apiserver uses for the default service. What does kubectl get ep output for you?

Comment (dlorenc) [inline pkg/localkube/localkube.go]: Something like this:
Unable to connect to the server: x509: certificate is valid for 10.0.2.15, 10.0.0.0, not 192.168.99.100

I think I have a fix ready, I'm just going to add all IPs on the machine as alternateIPs to the cert.

Comment (dlorenc) [inline pkg/localkube/localkube.go]: Ah whoops, looks like we're still missing the code to generate the CA and the client certificates...

Comment (luxas) [inline pkg/localkube/localkube.go]: Two files are created, apiserver.crt, which functions both as public key as well as ca. This cert should contain all ips as you suggested. The apiserver.key file serves as the private key. Not sure what you meant with the last comment...

Comment (dlorenc) [inline pkg/localkube/localkube.go]: So with easyrsa, we generated 5 certs:
ca.crt
kubecfg.crt	
kubecfg.key
kubernetes-master.crt
kubernetes-master.key

With this system we only have two. What certs should kubecfg be using to talk to the server? The same two?

**Pull request merged**